axum-macros/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+    clippy::mismatched_target_os,
+    clippy::await_holding_lock,
+    clippy::match_on_vec_items,
+    clippy::imprecise_flops,
+    clippy::suboptimal_flops,
+    clippy::lossy_float_literal,
+    clippy::rest_pat_in_fully_bound_structs,
+    clippy::fn_params_excessive_bools,
+    clippy::exit,
+    clippy::inefficient_to_string,
+    clippy::linkedlist,
+    clippy::macro_use_imports,
+    clippy::option_option,
+    clippy::verbose_file_reads,
+    clippy::unnested_or_patterns,
+    clippy::str_to_string,
+    rust_2018_idioms,
+    future_incompatible,
+    nonstandard_style,
+    missing_debug_implementations,
+    missing_docs
+)]
+#![deny(unreachable_pub, private_in_public)]
+#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
+#![forbid(unsafe_code)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(test, allow(clippy::float_cmp))]
+
+use quote::{quote, ToTokens};
+use syn::parse::Parse;
+
+mod from_request;
+
+/// Derive an implementation of [`FromRequest`].
+///
+/// Supports generating two kinds of implementations:
+/// 1. One that extracts each field individually.
+/// 2. Another that extracts the whole type at once via another extractor.
+///
+/// # Each field individually
+///
+/// By default `#[derive(FromRequest)]` will call `FromRequest::from_request` for each field:
+///
+/// ```
+/// use axum_macros::FromRequest;
+/// use axum::{
+///     extract::{Extension, TypedHeader},
+///     headers::ContentType,
+///     body::Bytes,
+/// };
+///
+/// #[derive(FromRequest)]
+/// struct MyExtractor {
+///     state: Extension<State>,
+///     content_type: TypedHeader<ContentType>,
+///     request_body: Bytes,
+/// }
+///
+/// #[derive(Clone)]
+/// struct State {
+///     // ...
+/// }
+///
+/// async fn handler(extractor: MyExtractor) {}
+/// ```
+///
+/// This requires that each field is an extractor (ie implements [`FromRequest`]).
+///
+/// ## Extracting via another extractor
+///
+/// You can use `#[from_request(via(...))]` to extract a field via another extractor, meaning the
+/// field itself doesn't need to implement `FromRequest`:
+///
+/// ```
+/// use axum_macros::FromRequest;
+/// use axum::{
+///     extract::{Extension, TypedHeader},
+///     headers::ContentType,
+///     body::Bytes,
+/// };
+///
+/// #[derive(FromRequest)]
+/// struct MyExtractor {
+///     // This will extracted via `Extension::<State>::from_request`
+///     #[from_request(via(Extension))]
+///     state: State,
+///     // and this via `TypedHeader::<ContentType>::from_request`
+///     #[from_request(via(TypedHeader))]
+///     content_type: ContentType,
+///     // Can still be combined with other extractors
+///     request_body: Bytes,
+/// }
+///
+/// #[derive(Clone)]
+/// struct State {
+///     // ...
+/// }
+///
+/// async fn handler(extractor: MyExtractor) {}
+/// ```
+///
+/// Note this requires the via extractor to be a generic tuple struct that implements `FromRequest`
+/// and has exactly one public field:
```

Comment:
Maybe use the term 'newtype struct'?

```suggestion
/// Note this requires the via extractor to be a generic newtype struct (a tuple struct with
/// exactly one public field) that implements `FromRequest`:
```